### ISSUE TITLE
linter: add a new rule that disallows an export named `then`

### ISSUE THREAD

**Issue by sapphi-red** [oxc-project/oxc#15795]
### Description

It would be nice to have a rule that disallows an export named `then`.

Object that has a `then` method (*thenables*) are treated similarly to Promises in JavaScript. When a module has an exported `then` function, that causes some strange behaviors. For example, the dynamic import and the static import will behave differently (https://github.com/tc39/proposal-dynamic-import/issues/47). This is also discouraged by MDN.

> **Warning**: Do not export a function called then() from a module. This will cause the module to behave differently when imported dynamically than when imported statically.
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import#module_namespace_object:~:text=Warning%3A%20Do%20not%20export%20a%20function%20called%20then()%20from%20a%20module.%20This%20will%20cause%20the%20module%20to%20behave%20differently%20when%20imported%20dynamically%20than%20when%20imported%20statically.

Especially, this becomes a problem when using Vitest, because Vitest internally converts static imports to dynamic imports (https://github.com/vitest-dev/vitest/issues/5122, https://github.com/vitest-dev/vitest/issues/4635).

The upcoming "import defer" proposal also have a special handling for `then` exports (does not have to be a function) ([Spec](https://tc39.es/proposal-defer-import-eval/#sec-IsSymbolLikeNamespaceKey)). Normally, accessing an export of the defer imported module triggers the module to be evaluated. But this is not the case when the export is named `then`.

It seems this lint was raised in https://github.com/eslint/eslint/issues/10428, but the implemented rule (`no-restricted-exports`) does not have the `then` export disallowed by default. I think we should set `then` as the default for that rule or add a new rule.

refs https://github.com/oxc-project/oxc/issues/479

### Examples
Examples of **incorrect** code for this rule:
```js
export let then

let then
export { then}

export function then() {}
```
Examples of **correct** code for this rule:
```js
let then

function then() {}
```

### Category
I guess this should be `correctness`. It is valid to use it, but should be discouraged.

**Comment by connorshea** [comment]
This has been an actual problem in libraries I've used in the past (e.g. https://github.com/ramda/ramda/issues/2751), I definitely would like to see this rule added
